module/validation/receipt_validator.go

```diff
@@ -148,29 +149,152 @@ func (v *receiptValidator) resultChainCheck(result *flow.ExecutionResult, prevRe
-	payloadExecutionResults := make(map[flow.Identifier]*flow.ExecutionResult)
-	for _, receipt := range receipts {
-		payloadExecutionResults[receipt.ExecutionResult.ID()] = &receipt.ExecutionResult
-	}
+	payloadExecutionResults := candidate.Payload.ResultsById()
 	// Build a functor that performs lookup first in receipts that were passed as payload and only then in
 	// local storage. This is needed to handle a case when same block payload contains receipts that
 	// reference each other.
 	// ATTENTION: Here we assume that ER is valid, this lookup can return a result which is actually invalid.
 	// Eventually invalid result will be detected and fail the whole validation.
-	previousResult := func(executionResult *flow.ExecutionResult) (*flow.ExecutionResult, error) {
-		prevResult, found := payloadExecutionResults[executionResult.PreviousResultID]
+	fetchResult := func(previousResultID flow.Identifier) (*flow.ExecutionResult, error) {
+		prevResult, found := payloadExecutionResults[previousResultID]
 		if found {
 			return prevResult, nil
 		}
 
-		return v.previousResult(executionResult)
+		return v.fetchResult(previousResultID)
+	}
+
+	header := candidate.Header
+	payload := candidate.Payload
+
+	// Get the latest sealed block on this fork, ie the highest block for which
+	// there is a seal in this fork. This block is not necessarily finalized.
+	last, err := v.seals.ByBlockID(header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve parent seal (%x): %w", header.ParentID, err)
 	}
+	sealed, err := v.headers.ByBlockID(last.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve sealed block (%x): %w", last.BlockID, err)
+	}
+	sealedHeight := sealed.Height
+
+	// forkBlocks is used to keep the IDs of the blocks we iterate through. We
+	// use it to identify receipts that are for blocks not in the fork.
+	forkBlocks := make(map[flow.Identifier]*flow.Header)
+
+	// Create a lookup table of all the receipts that are already included in
+	// blocks on the fork.
+	forkLookup := make(map[flow.Identifier]struct{})
+
+	// loop through the fork backwards, from parent to last sealed, and keep
+	// track of blocks and receipts visited on the way.
+	ancestorID := header.ParentID
+	for {
 
-	for i, r := range receipts {
-		err := v.validate(r, previousResult)
+		ancestor, err := v.headers.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor header (%x): %w", ancestorID, err)
+		}
+
+		// break out when we reach the sealed height
+		if ancestor.Height <= sealedHeight {
+			break
+		}
+
+		// keep track of blocks we iterate over
+		forkBlocks[ancestorID] = ancestor
+
+		// keep track of all receipts in ancestors
+		index, err := v.index.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor index (%x): %w", ancestorID, err)
+		}
+		for _, recID := range index.ReceiptIDs {
+			forkLookup[recID] = struct{}{}
+		}
+
+		ancestorID = ancestor.ParentID
+	}
```

Comment:
I've seen this pattern quite a lot, where we traverse through the chain from a block all the way to last sealed block, I think we could extract it and makes the logic here simplier:

```
var unsealedResultIDs []flow.Identifier
err := state.traverseUnsealedBlocksFrom(header.ParentID, func(ancestor *flow.Header) error {
		// keep track of blocks we iterate over
		forkBlocks[ancestorID] = ancestor

		// keep track of all receipts in ancestors
		index, err := v.index.ByBlockID(ancestorID)
		if err != nil {
			return fmt.Errorf("could not retrieve ancestor index (%x): %w", ancestorID, err)
		}
		for _, recID := range index.ReceiptIDs {
			forkLookup[recID] = struct{}{}
		}
		return nil
})
```

Note, we don't need to use `headers.ByBlockID`, we could just use protocol state, which [wraps the header.ByBlockID](https://github.com/onflow/flow-go/blob/master/state/protocol/badger/snapshot.go#L40)

Reply:
Put up a PR for it, please give your thoughts https://github.com/onflow/flow-go/pull/521